File: tools/cabana/dbc/dbcmanager.cc
#include "tools/cabana/dbc/dbcmanager.h"

#include <algorithm>
#include <numeric>

bool DBCManager::open(const SourceSet &sources, const QString &dbc_file_name, QString *error) {
  try {
    auto it = std::find_if(dbc_files.begin(), dbc_files.end(),
                           [&](auto &f) { return f.second && f.second->filename == dbc_file_name; });
    auto file = (it != dbc_files.end()) ? it->second : std::make_shared<DBCFile>(dbc_file_name);
    for (auto s : sources) {
      dbc_files[s] = file;
    }
  } catch (std::exception &e) {
    if (error) *error = e.what();
    return false;
  }

  emit DBCFileChanged();
  return true;
}

bool DBCManager::open(const SourceSet &sources, const QString &name, const QString &content, QString *error) {
  try {
    auto file = std::make_shared<DBCFile>(name, content);
    for (auto s : sources) {
      dbc_files[s] = file;
    }
  } catch (std::exception &e) {
    if (error) *error = e.what();
    return false;
  }

  emit DBCFileChanged();
  return true;
}

void DBCManager::close(const SourceSet &sources) {
  for (auto s : sources) {
    dbc_files[s] = nullptr;
  }
  emit DBCFileChanged();
}

void DBCManager::close(DBCFile *dbc_file) {
  for (auto &[_, f] : dbc_files) {
    if (f.get() == dbc_file) f = nullptr;
  }
  emit DBCFileChanged();
}

void DBCManager::closeAll() {
  dbc_files.clear();
  emit DBCFileChanged();
}

void DBCManager::addSignal(const MessageId &id, const cabana::Signal &sig) {
  if (auto m = msg(id)) {
    if (auto s = m->addSignal(sig)) {
      emit signalAdded(id, s);
      emit maskUpdated();
    }
  }
}

void DBCManager::updateSignal(const MessageId &id, const QString &sig_name, const cabana::Signal &sig) {
  if (auto m = msg(id)) {
    if (auto s = m->updateSignal(sig_name, sig)) {
      emit signalUpdated(s);
      emit maskUpdated();
    }
  }
}

void DBCManager::removeSignal(const MessageId &id, const QString &sig_name) {
  if (auto m = msg(id)) {
    if (auto s = m->sig(sig_name)) {
      emit signalRemoved(s);
      m->removeSignal(sig_name);
      emit maskUpdated();
    }
  }
}

void DBCManager::updateMsg(const MessageId &id, const QString &name, uint32_t size, const QString &node, const QString &comment) {
  auto dbc_file = findDBCFile(id);
  assert(dbc_file);  // This should be impossible
  dbc_file->updateMsg(id, name, size, node, comment);
  emit msgUpdated(id);
}

void DBCManager::removeMsg(const MessageId &id) {
  auto dbc_file = findDBCFile(id);
  assert(dbc_file);  // This should be impossible
  dbc_file->removeMsg(id);
  emit msgRemoved(id);
  emit maskUpdated();
}

QString DBCManager::newMsgName(const MessageId &id) {
  return QString("NEW_MSG_") + QString::number(id.address, 16).toUpper();
}

QString DBCManager::newSignalName(const MessageId &id) {
  auto m = msg(id);
  return m ? m->newSignalName() : "";
}

const std::vector<uint8_t> &DBCManager::mask(const MessageId &id) {
  static std::vector<uint8_t> empty_mask;
  auto m = msg(id);
  return m ? m->mask : empty_mask;
}

const std::map<uint32_t, cabana::Msg> &DBCManager::getMessages(uint8_t source) {
  static std::map<uint32_t, cabana::Msg> empty_msgs;
  auto dbc_file = findDBCFile(source);
  return dbc_file ? dbc_file->getMessages() : empty_msgs;
}

cabana::Msg *DBCManager::msg(const MessageId &id) {
  auto dbc_file = findDBCFile(id);
  return dbc_file ? dbc_file->msg(id) : nullptr;
}

cabana::Msg *DBCManager::msg(uint8_t source, const QString &name) {
  auto dbc_file = findDBCFile(source);
  return dbc_file ? dbc_file->msg(name) : nullptr;
}

QStringList DBCManager::signalNames() {
  // Used for autocompletion
  QStringList ret;
  for (auto &f : allDBCFiles()) {
    for (auto &[_, m] : f->getMessages()) {
      for (auto sig : m.getSignals()) {
        ret << sig->name;
      }
    }
  }
  ret.sort();
  ret.removeDuplicates();
  return ret;
}

int DBCManager::signalCount(const MessageId &id) {
  auto m = msg(id);
  return m ? m->sigs.size() : 0;
}

int DBCManager::signalCount() {
  auto files = allDBCFiles();
  return std::accumulate(files.cbegin(), files.cend(), 0, [](int &n, auto &f) { return n + f->signalCount(); });
}

int DBCManager::msgCount() {
  auto files = allDBCFiles();
  return std::accumulate(files.cbegin(), files.cend(), 0, [](int &n, auto &f) { return n + f->msgCount(); });
}

int DBCManager::dbcCount() {
  return allDBCFiles().size();
}

int DBCManager::nonEmptyDBCCount() {
  auto files = allDBCFiles();
  return std::count_if(files.cbegin(), files.cend(), [](auto &f) { return !f->isEmpty(); });
}

DBCFile *DBCManager::findDBCFile(const uint8_t source) {
  // Find DBC file that matches id.source, fall back to SOURCE_ALL if no specific DBC is found
  auto it = dbc_files.count(source) ? dbc_files.find(source) : dbc_files.find(-1);
  return it != dbc_files.end() ? it->second.get() : nullptr;
}

std::set<DBCFile *> DBCManager::allDBCFiles() {
  std::set<DBCFile *> files;
  for (const auto &[_, f] : dbc_files) {
    if (f) files.insert(f.get());
  }
  return files;
}

const SourceSet DBCManager::sources(const DBCFile *dbc_file) const {
  SourceSet sources;
  for (auto &[s, f] : dbc_files) {
    if (f.get() == dbc_file) sources.insert(s);
  }
  return sources;
}

QString toString(const SourceSet &ss) {
  return std::accumulate(ss.cbegin(), ss.cend(), QString(), [](QString str, int source) {
    if (!str.isEmpty()) str += ", ";
    return str + (source == -1 ? QStringLiteral("all") : QString::number(source));
  });
}

DBCManager *dbc() {
  static DBCManager dbc_manager(nullptr);
  return &dbc_manager;
}

File: tools/cabana/pycabana/dbc/dbcmanager.py
"""DBCManager - singleton for managing loaded DBC files."""

from PySide6.QtCore import QObject, Signal as QtSignal

from opendbc.can.dbc import DBC, Msg, Signal

from openpilot.tools.cabana.pycabana.dbc.dbc import MessageId


class DBCManager(QObject):
  """Singleton that manages loaded DBC files."""

  _instance: "DBCManager | None" = None

  # Signals
  dbcLoaded = QtSignal(str)  # dbc name
  signalAdded = QtSignal(MessageId, Signal)  # msg_id, signal
  signalRemoved = QtSignal(Signal)  # signal
  signalUpdated = QtSignal(Signal)  # signal
  msgUpdated = QtSignal(MessageId)  # msg_id
  msgRemoved = QtSignal(MessageId)  # msg_id
  DBCFileChanged = QtSignal()
  maskUpdated = QtSignal()

  def __init__(self, parent=None):
    super().__init__(parent)
    self._dbc: DBC | None = None
    self._name: str = ""

  @classmethod
  def instance(cls) -> "DBCManager":
    if cls._instance is None:
      cls._instance = DBCManager()
    return cls._instance

  def load(self, dbc_name: str) -> bool:
    """Load a DBC file by name."""
    try:
      self._dbc = DBC(dbc_name)
      self._name = dbc_name
      self.dbcLoaded.emit(dbc_name)
      return True
    except Exception as e:
      print(f"Failed to load DBC {dbc_name}: {e}")
      return False

  def save(self, filename: str) -> bool:
    """Save the DBC to a file."""
    if self._dbc is None:
      return False
    try:
      from openpilot.tools.cabana.pycabana.dbc.dbcfile import DBCFile
      dbc_file = DBCFile(self._dbc)
      dbc_string = dbc_file.to_dbc_string()
      with open(filename, 'w') as f:
        f.write(dbc_string)
      self._name = filename
      return True
    except Exception as e:
      print(f"Failed to save DBC to {filename}: {e}")
      return False

  def clear(self) -> None:
    """Clear the current DBC and create an empty one."""
    self._dbc = DBC("")
    self._name = ""
    self.DBCFileChanged.emit()

  @property
  def name(self) -> str:
    return self._name

  @property
  def dbc(self) -> DBC | None:
    return self._dbc

  def msg(self, msg_id: MessageId) -> Msg | None:
    """Get message definition by ID."""
    if self._dbc is None:
      return None
    return self._dbc.msgs.get(msg_id.address)

  def msgName(self, msg_id: MessageId) -> str:
    """Get message name, or empty string if unknown."""
    msg = self.msg(msg_id)
    return msg.name if msg else ""

  def addSignal(self, msg_id: MessageId, sig: Signal) -> None:
    """Add a signal to a message."""
    msg = self.msg(msg_id)
    if msg and self._dbc:
      msg.sigs[sig.name] = sig
      self.signalAdded.emit(msg_id, sig)
      self.maskUpdated.emit()

  def updateSignal(self, msg_id: MessageId, sig_name: str, new_sig: Signal) -> None:
    """Update a signal in a message."""
    msg = self.msg(msg_id)
    if msg and sig_name in msg.sigs:
      # If name changed, remove old and add new
      if sig_name != new_sig.name:
        del msg.sigs[sig_name]
        msg.sigs[new_sig.name] = new_sig
      else:
        msg.sigs[sig_name] = new_sig
      self.signalUpdated.emit(new_sig)
      self.maskUpdated.emit()

  def removeSignal(self, msg_id: MessageId, sig_name: str) -> None:
    """Remove a signal from a message."""
    msg = self.msg(msg_id)
    if msg and sig_name in msg.sigs:
      sig = msg.sigs[sig_name]
      self.signalRemoved.emit(sig)
      del msg.sigs[sig_name]
      self.maskUpdated.emit()

  def updateMsg(self, msg_id: MessageId, name: str, size: int, node: str, comment: str) -> None:
    """Update or create a message."""
    if self._dbc is None:
      return

    # Get existing message or create new one
    msg = self.msg(msg_id)
    if msg:
      # Update existing message
      msg.name = name
      msg.size = size
      if hasattr(msg, 'transmitter'):
        msg.transmitter = node
      if hasattr(msg, 'comment'):
        msg.comment = comment
    else:
      # Create new message
      new_msg = Msg(name=name, address=msg_id.address, size=size, sigs={})
      if hasattr(new_msg, 'transmitter'):
        new_msg.transmitter = node
      if hasattr(new_msg, 'comment'):
        new_msg.comment = comment
      self._dbc.msgs[msg_id.address] = new_msg

    self.msgUpdated.emit(msg_id)

  def removeMsg(self, msg_id: MessageId) -> None:
    """Remove a message from the DBC."""
    if self._dbc and msg_id.address in self._dbc.msgs:
      del self._dbc.msgs[msg_id.address]
      self.msgRemoved.emit(msg_id)
      self.maskUpdated.emit()

  def newMsgName(self, msg_id: MessageId) -> str:
    """Generate a new message name."""
    return f"NEW_MSG_{msg_id.address:X}"

  def newSignalName(self, msg_id: MessageId) -> str:
    """Generate a new signal name for a message."""
    msg = self.msg(msg_id)
    if not msg:
      return "NEW_SIGNAL_1"

    # Find the next available signal name
    i = 1
    while True:
      name = f"NEW_SIGNAL_{i}"
      if name not in msg.sigs:
        return name
      i += 1


def dbc_manager() -> DBCManager:
  """Global accessor for DBCManager singleton."""
  return DBCManager.instance()
